modules/reindex/src/main/java/org/elasticsearch/index/reindex/remote/RemoteRequestBuilders.java

```diff
@@ -193,45 +186,50 @@ private static String sortToUri(SortBuilder<?> sort) {
         throw new IllegalArgumentException("Unsupported sort [" + sort + "]");
     }
 
-    static String scrollPath() {
-        return "/_search/scroll";
-    }
+    static Request scroll(String scroll, TimeValue keepAlive, Version remoteVersion) {
+        Request request = new Request("POST", "/_search/scroll");
 
-    static Map<String, String> scrollParams(TimeValue keepAlive, Version remoteVersion) {
         if (remoteVersion.before(Version.V_5_0_0)) {
             /* Versions of Elasticsearch before 5.0 couldn't parse nanos or micros
              * so we toss out that resolution, rounding up so we shouldn't end up
              * with 0s. */
             keepAlive = timeValueMillis((long) Math.ceil(keepAlive.millisFrac()));
         }
-        return singletonMap("scroll", keepAlive.getStringRep());
-    }
+        request.addParameter("scroll", keepAlive.getStringRep());
 
-    static HttpEntity scrollEntity(String scroll, Version remoteVersion) {
         if (remoteVersion.before(Version.fromId(2000099))) {
             // Versions before 2.0.0 extract the plain scroll_id from the body
-            return new StringEntity(scroll, ContentType.TEXT_PLAIN);
+            request.setEntity(new StringEntity(scroll, ContentType.TEXT_PLAIN));
+            return request;
         }
+
         try (XContentBuilder entity = JsonXContent.contentBuilder()) {
-            return new StringEntity(Strings.toString(entity.startObject()
-                .field("scroll_id", scroll)
-                .endObject()), ContentType.APPLICATION_JSON);
+            entity.startObject()
+                    .field("scroll_id", scroll)
+                .endObject();
+            request.setEntity(new StringEntity(Strings.toString(entity), ContentType.APPLICATION_JSON));
```

Comment:
should we use `NStringEntity` instead? Which makes me wonder, should we make this easier for users, and expose   `setEntityAsString` and `setEntityAsByteArray` methods additionally to what we expose now??

Reply:
Hmmm. I think I'll merge as is and make a followup with those so we can talk about them!